### PR TITLE
Add support for selecting OpenGL/ES versions

### PR DESCRIPTION
**Pull request by MortimerGoro**
Add support for selecting OpenGL/ES versions. Required for WebGL 2.0 support in Servo ;)

### CODE REVIEW

File: src/platform/with_glx/native_gl_context.rs [hunk 1]
```diff
@@ -21,17 +24,52 @@ pub struct NativeGLContext {
...
+
+            let attributes = [
+                glx_extra::CONTEXT_MAJOR_VERSION_ARB as c_int, major as c_int,
+                glx_extra::CONTEXT_MINOR_VERSION_ARB as c_int, minor as c_int,
+                0
+            ];
+
+            // load the extra GLX functions
+            let extra_functions = glx_extra::Glx::load_with(|s| {
+                let c_str = CString::new(s.as_bytes().to_vec()).unwrap();
```
Comment (jdm) [inline src/platform/with_glx/native_gl_context.rs]: nit: no need for to_vec.

File: src/platform/with_glx/native_gl_context.rs [hunk 2]
```diff
@@ -21,17 +24,52 @@ pub struct NativeGLContext {
...
+                unsafe { glx::GetProcAddress(c_str.as_ptr() as *const u8) as *const _ }
+            });
+
+            unsafe {
+                extra_functions.CreateContextAttribsARB(display as * mut _,
```
Comment (jdm) [inline src/platform/with_glx/native_gl_context.rs]: nit: remove the space after *

File: src/platform/with_glx/native_gl_context.rs [hunk 3]
```diff
@@ -21,17 +24,52 @@ pub struct NativeGLContext {
...
+                                                        framebuffer_config,
+                                                        shared, 1 as glx::types::Bool,
+                                                        attributes.as_ptr())
+            }
+
```
Comment (jdm) [inline src/platform/with_glx/native_gl_context.rs]: nit: remove this extra newline.

File: src/platform/with_osmesa/mod.rs [hunk 4]
```diff
@@ -19,15 +21,36 @@ pub struct OSMesaContextHandle(osmesa_sys::OSMesaContext);
...
 
+        match *api_type {
+            gl::GlType::Gles => {
+                return Err("OpenGL ES is not supported");
+            },
+            _ => {}
+        }
+
+        let (major, minor) = match api_version {
+            GLVersion::Major(major) => { (major, 0) }, // OpenGL 2.1, 3.1
```
Comment (jdm) [inline src/platform/with_osmesa/mod.rs]: The comment doesn't match the code.

File: src/platform/with_osmesa/mod.rs [hunk 5]
```diff
@@ -19,15 +21,36 @@ pub struct OSMesaContextHandle(osmesa_sys::OSMesaContext);
...
+            GLVersion::MajorMinor(major, minor) => { (major, minor) }
```
Comment (jdm) [inline src/platform/with_osmesa/mod.rs]: The {}s are necessary for either match arm.

**Review by jdm — CHANGES_REQUESTED** [review]
It's amazing how many places need to pass around the api version now.

**Comment by MortimerGoro** [comment]
`It's amazing how many places need to pass around the api version now.`

Yes, that's difficult to refactor because everything is initialized in the constructor.  Maybe we can use a more generic struct to unify all constructor parameters in the future and reduce the number of parameters passed to utility functions.

**Pull request merged**